### PR TITLE
[DOC] Enhanced RDoc for Regexp

### PR DESCRIPTION
**Pull request by BurdetteLamar**
Treats:
- ::new
- ::escape
- ::try_convert
- ::union
- ::last_match

### CODE REVIEW

File: re.c [hunk 1]
```diff
@@ -3540,28 +3539,47 @@ rb_reg_match_p(VALUE re, VALUE str, long pos)
...
+ *    Regexp.new(regexp) -> regexp
+ *
+ *  With argument +string+ given, returns a new regexp with the given string
+ *  and options:
+ *
+ *    r = Regexp.new('foo') # => /foo/
+ *    r.source              # => "foo"
+ *    r.options             # => 0
+ *
+ *  The +string+ may contain interpolations:
```
Comment (jeremyevans) [inline re.c:3552]: I don't think it's important to mention this or give an example, as any string may contain interpolations.  the string may also be a heredoc, or a non-literal string.

File: re.c [hunk 2]
```diff
@@ -3540,28 +3539,47 @@ rb_reg_match_p(VALUE re, VALUE str, long pos)
...
+ *  - The logical OR of one or more of the constants
+ *    Regexp::EXTENDED, Regexp::IGNORECASE, and Regexp::MULTILINE:
+ *
+ *      Regexp.new('foo', Regexp::IGNORECASE) # => /foo/i
+ *      Regexp.new('foo', Regexp::EXTENDED)   # => /foo/x
+ *      Regexp.new('foo', Regexp::MULTILINE)  # => /foo/m
+ *      flags = Regexp::IGNORECASE | Regexp::EXTENDED |  Regexp::MULTILINE
+ *      Regexp.new('foo', flags)              # => /foo/mix
+ *
+ *  - +nil+ or +false+, which is ignored.
```
Comment (jeremyevans) [inline re.c:3568]: I would not document or give an example with a non-numeric value for the second argument.

File: re.c [hunk 3]
```diff
@@ -3764,19 +3786,20 @@ rb_check_regexp_type(VALUE re)
...
+ *    Regexp.try_convert(object) -> regexp or nil
+ *
+ *  Returns +object+ if it is a regexp:
+ *
+ *    Regexp.try_convert(/re/) # => /re/
  *
- *  Try to convert <i>obj</i> into a Regexp, using to_regexp method.
- *  Returns converted regexp or nil if <i>obj</i> cannot be converted
- *  for any reason.
+ *  Otherwise if +object+ responds to <tt>:to_hash</tt>,
```
Comment (jeremyevans) [inline re.c:3795]: All of the `to_hash` in this method documentation needs to be switched to `to_regexp`.

**Review by jeremyevans — CHANGES_REQUESTED** [review]
Looks mostly good.  A few changes requested.

File: re.c (hunk 1 above)
Comment (BurdetteLamar) [reply re.c]: Removed.

File: re.c (hunk 2 above)
Comment (BurdetteLamar) [reply re.c:3563]: Removed.

File: re.c (hunk 3 above)
Comment (BurdetteLamar) [reply re.c]: Fixed.

**Review by jeremyevans — APPROVED** [review]

**Pull request merged**